cg.go

```diff
@@ -0,0 +1,293 @@
+package optimize
+
+import (
+	"math"
+
+	"github.com/gonum/floats"
+)
+
+const (
+	iterationRestartFactor = 6
+	angleRestartThreshold  = 0.9
+)
+
+// CGVariant represents the formula for computing the scaling parameter (in
+// literature usually denoted by the Greek letter β) used for updating the
+// conjugate direction in the nonlinear conjugate gradient (CG) method.
+type CGVariant interface {
+	// Init is called at the first iteration and provides a way to initialize
+	// any internal state.
+	Init(loc Location)
+	// Beta returns the value of the scaling parameter that is computed
+	// according to the particular variant of the CG method.
+	Beta(grad, gradPrev, dirPrev []float64) float64
+}
+
+// CG implements the nonlinear conjugate gradient method for solving nonlinear
+// unconstrained optimization problems [1]. It is a line search method that
+// generates the search directions d_k according to the formula
+// d_{k+1} = -∇f_{k+1} + β_k*d_k,   d_0 = -∇f_0.
+// Different variants of the conjugate gradient method differ in the choice
+// of the parameter β_k. Conjugate gradient method requires no matrix storage
+// and is faster than the gradient descent method, but L-BFGS is usually more
+// efficient and robust.
+//
+// CG implements a restart strategy that takes the steepest descent direction
+// (i.e., d_{k+1} = -∇f_{k+1}) whenever any of the following conditions holds.
+// * A certain number of iterations has elapsed without a restart. This number
```

Comment:
They come through as asterisks just fine.  Also, the ArmijoConditionMet function uses a different format for bullet points ("-" instead of "*") and also has incorrect spacing in its fixed width section.  Several of the comments have different fixed width formatting sections.

You should be able to see it at http://godoc.org/github.com/jonlawlor/optimize#CG